feat: 유저 도메인 RabbitMQ 장애 대응 시스템 구현

Mybread2 opened the pull request:
### 요청 흐름
1. 클라이언트가 회원탈퇴 요청
2. 아웃박스 테이블에 이벤트가 저장되고 스프링 이벤트 발행
3. RabbitMQ exchange에 메시지 전송하고
4. 성공하면 이벤트 리스너에서 아웃박스 테이블에 저장된 특정 사용자의 이벤트 타입을 발행 완료로 변경
5. 만약 실패한다면 5분 주기로 published = 0인 컬럼들을 재시도
6. 새벽 3시에 published = 1인, 메시지 전송에 성공한 이벤트들을 삭제

### DDL 추가 (유저 아웃박스 테이블)
-- 사용자 아웃박스 이벤트 테이블
```
CREATE TABLE user_outbox_events
(
    id            BIGINT AUTO_INCREMENT PRIMARY KEY,
    user_id       BIGINT       NOT NULL,
    event_type    VARCHAR(100) NOT NULL,
    event_data    TEXT,
    published     BOOLEAN      NOT NULL DEFAULT FALSE,
    published_at  TIMESTAMP    NULL,
    last_retry_at TIMESTAMP    NULL,
    created_at    TIMESTAMP    NOT NULL DEFAULT CURRENT_TIMESTAMP
);
```

### 예상 실패 케이스
1. 네트워크/연결 문제
2. Exchange가 존재하지 않음
3. 라우팅 실패
4. 메시지 직렬화 실패
5. 메모리/디스크 부족
6. 타임아웃

여기서 보통 발생할 문제는 1번인데, 사실 RabbitMQ 서버를 재가동하거나 용량을 늘리는 작업은 베포하는 과정에서 설계를 잘 해놓으면 되는 문제같습니다. 2~6번 상황은 코드만 제대로 구현 되어있으면 발생하지 않는 문제고 실제 운영 단계에서는 사실상 RabbitMQ 서버 자체의 문제일 가능성이 크다고 생각합니다.

### 왜 재시도 최대 카운트를 삭제했나
  주로 레빗MQ 서버에 문제가 생겼을 때만 거의 메시지 전송에 실패를 할텐데, 실패한 것에 대해서 재시도 카운트가 존재를 하면 계속해서 실패하는 이벤트를 재시도를 할 때 재시도카운트를 업데이트를 해줘야하기 때문에 이는 DB에 영향이 갈거같다고 생각했습니다. 
  그래서 재시도 카운트를 없앰으로써 계속해서 재시도를 하고 계속 실패를 하더라도 단순히 조회만 하기때문에 DB에 영향을 상대적으로 덜 준다고 판단했습니다.

Mybread2 committed:
refactor: RabbitMQ 상수를 global 패키지로 이동

- user 도메인의 RabbitMQ 상수들을 global/rabbitmq/constant로 이동
- UserEventPublisher에서 global 상수 사용하도록 수정 및 라우팅 키 버그 수정
- 메시징 설정 중앙화로 코드 유지보수성 향상

Mybread2 committed:
refactor: producer 패키지 구조 분리 - 아웃박스 패턴 준비

- 역할별 패키지 분리로 책임 명확화

Mybread2 committed:
feat: 아웃박스 성공 이벤트들을 스케쥴링할 클래스 생성

Mybread2 committed:
feat: 회원탈퇴 아웃박스 패턴 구현

- UserOutboxEvent 엔티티 추가 (아웃박스 이벤트 저장)
- UserOutboxEventRepository 인터페이스 및 구현체 추가
- UserOutboxService 비즈니스 로직 계층 추가
- UserEvents.Withdrawn 스프링 이벤트 추가
- UserEventHandler 트랜잭션 커밋 후 RabbitMQ 발행 처리
- UserOutboxCleanupScheduler 실패 재시도 및 정리 스케줄러 추가

데이터 일관성 보장을 위한 아웃박스 패턴 적용:
1. 회원탈퇴 → 아웃박스 저장 + 스프링 이벤트 발행
2. 트랜잭션 커밋 후 → RabbitMQ 발행 + 아웃박스 상태 업데이트
3. 스케줄러를 통한 실패 이벤트 재시도 및 오래된 데이터 정리

Mybread2 committed:
refactor: 불필요한 클래스 및 메서드 제거
- 지금 당장은 회원탈퇴에 대해서만 이벤트를 발행하면되고, 이벤트를 발행하게 되면 계속해서 exchange에 쌓이고 아웃박스 테이블에도 쌓이기 때문에 추후에 필요하다면 구현할 예정

Mybread2 committed:
Merge branch 'dev-v3' of https://github.com/momo-spring-project/momo-backend into feat/AWS-v3

Mybread2 committed:
refactor: UserOutboxEventRepository 수정

- save() 메서드를 UserOutboxJpaRepository에 위임하여 기본 CRUD 활용
- 복잡한 JPQL 쿼리들은 UserOutboxQueryRepository(Querydsl)로 분리하여 구현
- UserOutboxEventRepositoryImpl에서 두 리포지토리를 주입받아 역할별 메서드 위임 처리
- 코드 가독성 및 유지보수성 향상
- 결제 아웃박스 테이블이 V12로 올라올 예정이라 V13으로 생성

Mybread2 committed:
refactor: UserOutboxEventRepository 수정

- save() 메서드를 UserOutboxJpaRepository에 위임하여 기본 CRUD 활용
- 복잡한 JPQL 쿼리들은 UserOutboxQueryRepository(Querydsl)로 분리하여 구현
- UserOutboxEventRepositoryImpl에서 두 리포지토리를 주입받아 역할별 메서드 위임 처리
- 코드 가독성 및 유지보수성 향상
- 결제 아웃박스 테이블이 V12로 올라올 예정이라 V13으로 생성

Mybread2 committed:
fix: @TransactionalEventListener에서 새로운 트랜잭션으로 아웃박스 상태 업데이트

- UserOutboxServiceImpl.markEventAsPublished()에 REQUIRES_NEW 전파 속성 적용
- AFTER_COMMIT 단계에서 호출되는 메서드의 TransactionRequiredException 해결
- 아웃박스 패턴의 메시지 발행 후 상태 업데이트 정상화

Mybread2 committed:
refactor: 아웃박스 이벤트 재시도 로직 단순화를 위한 retryCount 제거

- RabbitMQ 장애 시에만 메시지 발송이 실패하는 점을 고려해 retryCount 컬럼 제거
- 재시도 실패 이벤트에 대해 별도 카운팅 없이 계속 재시도되도록 로직 수정
- retryCount 증가 로직 및 관련 조건(query, scheduler) 제거
- 조회만 수행하므로 DB 부하 감소 효과 기대

Mybread2 committed:
refactor: 유저 아웃박스 테이블 DDL V13 에서 V12로 수정

leeuihyun commented on src/main/java/com/example/momo/global/rabbitmq/constant/RoutingKeys.java:
```diff
@@ -1,4 +1,6 @@
 package com.example.momo.global.rabbitmq.constant;
 
 public class RoutingKeys {
+	// User 관련 routing key들
+	public static final String USER_WITHDRAWN = "user.withdrawn";
```
아 이런식으로 관리한다고 적어놓으셨군요.. 이부분을 못봤었네요 확인했습니다.
병합되면 아까 올렸던 PR 수정해서 병합하겠습니다

Mybread2 replied:
확인했습니다!!

leeuihyun commented on src/main/java/com/example/momo/global/rabbitmq/dto/UserEventMessage.java:
```diff
@@ -30,40 +29,6 @@ public record UserWithdrawnData(
 	) {
 	}
 
-	// 회원가입 이벤트 데이터
-	public record UserRegisteredData(
-		Long userId,
-		String nickname,
-		String email,
-		Double latitude,
-		Double longitude,
-		List<Integer> categoryIds,
-		LocalDateTime registeredAt
-	) {
-	}
-
-	// 팔로우 이벤트 데이터
-	public record UserFollowedData(
-		Long followerId,
-		Long followingId,
-		String followerNickname,
-		String followingNickname,
-		LocalDateTime followedAt
-	) {
-	}
-
-	// 사용자 평가 이벤트 데이터
-	public record UserRatedData(
-		Long reviewerId,
-		Long targetUserId,
-		Long meetingId,
-		Integer ratingScore,
-		String reviewerNickname,
-		String targetUserNickname,
-		LocalDateTime ratedAt
-	) {
-	}
```
이 부분 삭제 이유는 뭔가요? 미사용에 의한 삭제인가요?

Mybread2 replied:
맞습니다. 우선 사용될만한 이벤트들을 다 발행을 해놨었는데, 이게 Queue가 없든 있든 지속적으로 exchange에 접근을 하기 때문에 지금 당장은 불필요할거같아서 삭제했습니다. 추후에 필요한 도메인이 있다면 금방 추가할 수 있도록 틀은 잡아놨습니다!

leeuihyun replied:
아 넵 확인했습니다

leeuihyun commented on src/main/java/com/example/momo/domain/user/event/springEvent/UserEventHandler.java:
```diff
@@ -0,0 +1,45 @@
+	private final UserEventPublisher userEventPublisher;
+	private final UserOutboxService userOutboxService;
+
+	@TransactionalEventListener(phase = TransactionPhase.AFTER_COMMIT)
```
아웃박스 패턴을 적용하기 위해서 SpringEvent를 추가로 사용하신게 맞나요?

Mybread2 replied:
맞습니다. 아웃박스 패턴만을 위한 SpringEvent입니다.

leeuihyun reviewed (changes requested):
고생하셨습니다.

아웃박스 패턴이 확실히 안정성이 더 높겠네요. 아무래도 rabbitMQ 또한 서버다운되었을 때는 dlq를 통해 처리하지 못하는 부분이 있을 것 같습니다.

Mybread2 committed:
fix: 메시지 발행 실패 시 published 상태가 잘못 업데이트되는 문제 수정

- RabbitMQ 메시지 발행 실패 시 예외를 던지지 않아, published=true로 마킹되는 문제 발생
- publishEvent 내부에서 Exception 발생 시 RuntimeException을 다시 던지도록 수정
- 메시지 발행 성공한 경우에만 Outbox 이벤트를 발행 완료 상태로 업데이트되도록 보장

hojin915 reviewed (approved):
고생하셨습니다

제가 말씀드렸던 아웃박스 캐싱에 대해 더 찾아봤는데 DB보다 더 불안정한 캐시를 사용하고 캐시를 기준으로 폴링을 하는 것은 메세지 전송 보장이라는 아웃박스 패턴의 목적과 맞지 않을 수 있다고 합니다.
